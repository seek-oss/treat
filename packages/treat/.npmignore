node_modules
src
tests
tsconfig.json
*.tsbuildinfo