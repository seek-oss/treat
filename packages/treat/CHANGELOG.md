# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [1.5.1](https://github.com/seek-oss/treat/tree/master/packages/treat/compare/v1.5.0...v1.5.1) (2020-07-27)


### Bug Fixes

* Allow custom treat file module rules ([#128](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/128)) ([a0f1bad](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/a0f1bad))





# [1.5.0](https://github.com/seek-oss/treat/tree/master/packages/treat/compare/v1.4.3...v1.5.0) (2020-07-23)


### Features

* Add hot module reloading support ([#127](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/127)) ([ce5b667](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/ce5b667))





## [1.4.3](https://github.com/seek-oss/treat/tree/master/packages/treat/compare/v1.4.2...v1.4.3) (2020-07-13)


### Bug Fixes

* Ensure consistent and valid content hashes ([#111](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/111)) ([d6993a4](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/d6993a4)), closes [#121](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/121)





## [1.4.1](https://github.com/seek-oss/treat/tree/master/packages/treat/compare/v1.4.0...v1.4.1) (2020-04-26)


### Bug Fixes

* **webpack-plugin:** Checking treat file extension ([#107](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/107)) ([5f5e838](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/5f5e838))





## [1.2.4](https://github.com/seek-oss/treat/tree/master/packages/treat/compare/v1.2.3...v1.2.4) (2020-01-31)


### Bug Fixes

* **styleMap:** Support numbers as keys on 'styleMap' objects ([#88](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/88)) ([1fd5f1f](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/1fd5f1f))





## [1.2.2](https://github.com/seek-oss/treat/tree/master/packages/treat/compare/v1.2.1...v1.2.2) (2020-01-15)


### Bug Fixes

* Detect multiple webpack versions error ([#80](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/80)) ([c067bdf](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/c067bdf))





## [1.2.1](https://github.com/seek-oss/treat/tree/master/packages/treat/compare/v1.2.0...v1.2.1) (2020-01-12)


### Bug Fixes

* Handle undefined watcher error ([#82](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/82)) ([c3b44e9](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/c3b44e9))





## [1.1.7](https://github.com/seek-oss/treat/tree/master/packages/treat/compare/v1.1.6...v1.1.7) (2019-12-02)


### Bug Fixes

* Produce single type definitions ([#70](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/70)) ([b60aebd](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/b60aebd))





## [1.1.6](https://github.com/seek-oss/treat/tree/master/packages/treat/compare/v1.1.5...v1.1.6) (2019-10-20)


### Bug Fixes

* Handle zeroes in treat class selectors ([#66](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/66)) ([2d44b72](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/2d44b72))





## [1.1.5](https://github.com/seek-oss/treat/tree/master/packages/treat/compare/v1.1.4...v1.1.5) (2019-10-15)


### Bug Fixes

* Honour input order for simple pseudos ([#65](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/65)) ([ee7ae3e](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/ee7ae3e))





## [1.1.4](https://github.com/seek-oss/treat/tree/master/packages/treat/compare/v1.1.3...v1.1.4) (2019-10-08)


### Bug Fixes

* Gracefully handle compilation errors ([#64](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/64)) ([47dd4e8](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/47dd4e8))





## [1.1.3](https://github.com/seek-oss/treat/tree/master/packages/treat/compare/v1.1.2...v1.1.3) (2019-09-20)


### Bug Fixes

* Normalize CSS file paths for Windows ([#61](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/61)) ([0c96664](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/0c96664))





## [1.1.2](https://github.com/seek-oss/treat/tree/master/packages/treat/compare/v1.1.1...v1.1.2) (2019-09-20)


### Bug Fixes

* Add theme.d.ts to npm package ([#59](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/59)) ([39a7c63](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/39a7c63))





## [1.1.1](https://github.com/seek-oss/treat/tree/master/packages/treat/compare/v1.1.0...v1.1.1) (2019-09-19)


### Bug Fixes

* Handle tree shaking of themes ([#60](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/60)) ([26ebcd8](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/26ebcd8))





## [1.0.3](https://github.com/seek-oss/treat/tree/master/packages/treat/compare/v1.0.2...v1.0.3) (2019-08-01)


### Bug Fixes

* Fail validation for undefined styles ([#55](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/55)) ([b8399d9](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/b8399d9))





## [1.0.2](https://github.com/seek-oss/treat/tree/master/packages/treat/compare/v1.0.1...v1.0.2) (2019-08-01)


### Bug Fixes

* **deps:** Update Autoprefixer, suppresses browserslist warning ([#54](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/54)) ([dde0281](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/dde0281))





# [1.0.0](https://github.com/seek-oss/treat/tree/master/packages/treat/compare/v1.0.0-beta.2...v1.0.0) (2019-07-24)


### Bug Fixes

* Correctly order themed styles from multiple modules ([#2](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/2)) ([8393fed](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/8393fed))
* Don't publish global theme declaration ([#5](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/5)) ([4e141cc](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/4e141cc))
* Dont throw selector error until all themes found ([#12](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/12)) ([5c01c96](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/5c01c96))
* Handle themed OR selectors in local styles ([#20](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/20)) ([f5e5049](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/f5e5049))
* Make ‘resolveStyles’ support deep objects, add ‘resolveClassName’ ([#6](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/6)) ([a94d920](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/a94d920))
* Mark all treat files as having side effects ([#16](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/16)) ([828a1dd](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/828a1dd))
* **globalStyle:** Handle selectors correctly when no local styles ([#11](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/11)) ([09d80d0](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/09d80d0))
* Mark loader as cacheable ([#23](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/23)) ([53cd2dd](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/53cd2dd))
* Recompile styles when theme dependencies update ([#26](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/26)) ([2f03b8e](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/2f03b8e))
* Remove css dependency when tree shaking ([#4](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/4)) ([114fcf6](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/114fcf6))
* Rename `css` function to `styleMap` ([#7](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/7)) ([110848b](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/110848b))
* **globalStyle:** Correct type signature ([#15](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/15)) ([1b7d025](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/1b7d025))
* **keyframes:** Handle keyframes in selectors block ([#10](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/10)) ([f973d5b](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/f973d5b))
* **resolveClassNames:** Handle theme dereferencing in object keys ([#3](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/3)) ([7957554](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/7957554))
* **selectors:** Enforce targeting of ‘&’ in selectors ([#17](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/17)) ([728c9fa](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/728c9fa))
* **styleTree:** Remove possible styleTree class clashes ([#34](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/34)) ([64862eb](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/64862eb))
* Simplify globalStyles type ([#14](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/14)) ([791fb89](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/791fb89))
* Support Safari < 9 by using var for named exports ([#37](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/37)) ([fc36761](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/fc36761))


### Features

* Add 'styleTree' function ([#28](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/28)) ([88ff843](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/88ff843))
* Add [@keyframes](https://github.com/keyframes) API ([#8](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/8)) ([f115e78](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/f115e78))
* Add runtime validation of styles, optimize compilation ([#18](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/18)) ([5018046](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/5018046))
* Add webpack-plugin option validation ([#32](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/32)) ([a40e7aa](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/a40e7aa))
* Allow ‘undefined’ to be exported from treat files ([#30](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/30)) ([d302abd](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/d302abd))
* Allow themeable global styles ([#19](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/19)) ([1da4b89](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/1da4b89))
* Release v1.0.0 ([#39](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/39)) ([ee7e890](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/ee7e890))


### Performance Improvements

* Add compiler cache ([#25](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/25)) ([616627a](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/616627a))
* Avoid unnecessary rebuilds ([#24](https://github.com/seek-oss/treat/tree/master/packages/treat/issues/24)) ([64c994e](https://github.com/seek-oss/treat/tree/master/packages/treat/commit/64c994e))
