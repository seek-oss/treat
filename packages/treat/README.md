# 🍬 treat

## [Docs](https://seek-oss.github.io/treat/)

See the documentation at seek-oss.github.io/treat for more information about using treat.

## License

MIT.
