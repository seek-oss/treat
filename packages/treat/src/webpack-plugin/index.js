import Promise from 'bluebird';
import partition from 'lodash/partition';
import chalk from 'chalk';

import store from './store';
import reIndexModules from './reIndexModules';
import TreatError from './TreatError';
import makeTreatCompiler from './treatCompiler';
import optionValidator from './optionValidator';
import { debug } from './utils';
import createCompat from './compat';

const isProductionLikeMode = (options) => {
  return options.mode === 'production' || !options.mode;
};

const trace = debug('treat:webpack-plugin');

const TWP = 'treat-webpack-plugin';

const makeOptionDefaulter = (prodLike) => (option, { dev, prod }) => {
  if (typeof option !== 'undefined') {
    return option;
  }

  return prodLike ? prod : dev;
};

export class TreatPlugin {
  constructor(options = {}) {
    optionValidator(options);

    const {
      test = /\.treat\.(js|ts)$/,
      outputCSS = true,
      outputLoaders = ['style-loader'],
      localIdentName,
      themeIdentName,
      minify,
      browsers,
      hmr,
    } = options;

    this.store = store();
    this.treatCompiler = makeTreatCompiler();

    this.test = test;
    this.minify = minify;
    this.localIdentName = localIdentName;
    this.themeIdentName = themeIdentName;
    this.loaderOptions = {
      outputCSS,
      outputLoaders,
      browsers,
      hmr,
    };
  }

  apply(compiler) {
    const isWebpack5 = Boolean(compiler.webpack && compiler.webpack.version);
    const compat = createCompat(isWebpack5);

    compiler.hooks.watchRun.tap(TWP, (watchCompiler) => {
      const modifiedFiles = compat.getModifiedFiles(watchCompiler);

      this.treatCompiler.expireCache(modifiedFiles);
    });

    compiler.hooks.thisCompilation.tap(TWP, (compilation) => {
      let allCssModules = [];
      let usedCssModules;

      const rebuildModule = (module) =>
        new Promise((resolve) => {
          compilation.rebuildModule(module, () => {
            resolve();
          });
        });

      compilation.hooks.finishModules.tapPromise(TWP, async () => {
        // Filter out css modules that no longer exist (watch mode)
        this.store
          .getAllCssOwners()
          .filter((identifier) => !Boolean(compilation.findModule(identifier)))
          .forEach((identifier) => {
            this.store.deleteCssRequests(identifier);
          });

        // Filter out theme treat modules that no longer exist (watch mode)
        this.store
          .getThemeIdentifiers()
          .filter((identifier) => !Boolean(compilation.findModule(identifier)))
          .forEach((identifier) => {
            this.store.deleteModuleThemes(identifier);
          });

        const finalThemeHash = this.store.getThemesHash();
        const builtTreatModules = this.store
          .popAllModules()
          .filter(
            ({ moduleIdentifier }) =>
              !this.store.getThemeIdentifiers().includes(moduleIdentifier),
          );

        // Rebuild all non theme treat modules that built before all themes were registered
        const staleModules = builtTreatModules.filter(
          ({ themeHash }) => themeHash != finalThemeHash,
        );

        await Promise.map(staleModules, async ({ moduleIdentifier }) => {
          trace('Rebuiling stale module: %i', moduleIdentifier);
          const treatModule = compilation.findModule(moduleIdentifier);

          if (!treatModule) {
            // In cases where the build fails, sometimes modules are not present
            // on the compilation object. Skip rebuilding in these cases.
            trace(
              "Bail on rebuild of stale module: %i Module doesn't exist on compilation",
              moduleIdentifier,
            );
            return;
          }

          await rebuildModule(treatModule);
        });

        // Modules should be upto date so clear rebuild queue
        this.store.popAllModules();

        if (builtTreatModules.length > 0) {
          // Rebuild all theme modules to ensure they contain upto date css
          await Promise.map(
            this.store.getThemeIdentifiers(),
            async (themeModuleIdentifier) => {
              trace('Rebuiling theme module: %i', themeModuleIdentifier);

              const treatModule = compilation.findModule(themeModuleIdentifier);

              if (!treatModule) {
                // In cases where the build fails, sometimes modules are not present
                // on the compilation object. Skip rebuilding in these cases.
                trace(
                  "Bail on rebuild of theme module: %i Module doesn't exist on compilation",
                  themeModuleIdentifier,
                );
                return;
              }

              await rebuildModule(treatModule);
            },
          );
        }

        const cssResources = this.store
          .getAllThemedCssRequests()
          .map(({ resource }) => resource);

        const getTreatCssResource = (module) => {
          if (
            module.matchResource &&
            cssResources.includes(module.matchResource)
          ) {
            return module.matchResource;
          }

          if (module.type === 'css/mini-extract') {
            const issuer = compat.getModuleIssuer(compilation, module);

            return getTreatCssResource(issuer);
          }

          return null;
        };

        for (const module of compilation._modules.values()) {
          const treatCssResource = getTreatCssResource(module);

          if (treatCssResource) {
            const { owner, type, theme } = this.store.getThemedCssModuleInfo(
              treatCssResource,
            );
            const themeIdentifier = theme
              ? this.store.getTheme(theme).identifier
              : null;

            allCssModules.push({
              type,
              module: compilation.findModule(module.identifier()),
              owner: compilation.findModule(owner),
              themeModule: theme
                ? compilation.findModule(themeIdentifier)
                : null,
              identifier: module.identifier(),
            });
          }
        }
      });

      compilation.hooks.optimizeDependencies.tap(
        // Running during advanced stage means export usage info has been added to the graph already
        { name: TWP, stage: 10 },
        () => {
          // Some modules may not be used and their css can be safely be removed from the chunks
          const cssModuleGroups = partition(
            allCssModules,
            ({ module, owner, themeModule }) => {
              const ownerIsUsed = compat.isModuleUsed(compilation, owner);
              const themeIsUsed =
                !themeModule || compat.isModuleUsed(compilation, themeModule);

              const cssModuleIsUsed = ownerIsUsed && themeIsUsed;

              if (!cssModuleIsUsed) {
                trace(
                  'CSS Module marked for removal due to unused %s %m',
                  chalk.yellow(ownerIsUsed ? 'theme' : 'owner'),
                  module,
                );
              }

              return cssModuleIsUsed;
            },
          );

          usedCssModules = cssModuleGroups[0];
          const cssModulesToRemove = cssModuleGroups[1];

          if (cssModulesToRemove.length > 0) {
            const themeModules = this.store
              .getThemeIdentifiers()
              .map((ident) => compilation.findModule(ident));

            const cssModulesIdentsToRemove = cssModulesToRemove.map(
              ({ identifier }) => identifier,
            );

            for (const themeModule of themeModules) {
              const depsToRemove = themeModule.dependencies.filter((dep) => {
                const depModule = compat.getDependencyModule(compilation, dep);

                const shouldRemove =
                  depModule &&
                  cssModulesIdentsToRemove.includes(depModule.identifier());

                if (shouldRemove) {
                  trace('Mark %m for clean-up in %m', depModule, themeModule);
                }

                return shouldRemove;
              });

              for (const dep of depsToRemove) {
                themeModule.removeDependency(dep);
              }

              compilation.removeReasonsOfDependencyBlock(themeModule, {
                dependencies: depsToRemove,
                blocks: [],
              });
            }
          }
        },
      );

      compilation.hooks.afterChunks.tap(TWP, (chunks) => {
        try {
          // afterChunks hook means module/chunk order properties have been set
          // We can now correct those values by referencing the ordering of the owner treat file
          for (const chunk of chunks) {
            const cssModulesInChunk = usedCssModules.filter(({ module }) =>
              compat.isModuleInChunk(compilation, module, chunk),
            );

            for (const chunkGroup of chunk.groupsIterable) {
              // Corrects ChunkGroup._moduleIndicies
              // Used for mini-extract-css-plugin ordering
              reIndexModules(
                cssModulesInChunk,
                {
                  getPreIndex: ({ module }) =>
                    compat.getCGModulePreOrderIndex(chunkGroup, module),
                  getPostIndex: ({ module }) =>
                    compat.getCGModulePostOrderIndex(chunkGroup, module),
                  getOwnerIndex: ({ owner }) =>
                    compat.getModulePreOrderIndex(compilation, owner),
                  getThemeIndex: ({ themeModule }) =>
                    compat.getModulePreOrderIndex(compilation, themeModule),
                  setPreIndex: ({ module }, i) =>
                    compat.setCGModulePreOrderIndex(chunkGroup, module, i),
                  setPostIndex: ({ module }, i) =>
                    compat.setCGModulePostOrderIndex(chunkGroup, module, i),
                },
                'ChunkGroup moduleIndicies',
              );
            }
          }

          // Corrects Module.index/index2
          reIndexModules(
            usedCssModules,
            {
              getPreIndex: ({ module }) =>
                compat.getModulePreOrderIndex(compilation, module),
              getPostIndex: ({ module }) =>
                compat.getModulePostOrderIndex(compilation, module),
              getOwnerIndex: ({ owner }) =>
                compat.getModulePreOrderIndex(compilation, owner),
              getThemeIndex: ({ themeModule }) =>
                compat.getModulePreOrderIndex(compilation, themeModule),
              setPreIndex: ({ module }, i) =>
                compat.setModulePreOrderIndex(compilation, module, i),
              setPostIndex: ({ module }, i) =>
                compat.setModulePostOrderIndex(compilation, module, i),
            },
            'Module.index/index2',
          );

          this.store
            .getThemeIdentifiers()
            .map((themeModuleIdentifier) =>
              compilation.findModule(themeModuleIdentifier),
            )
            .forEach((themeModule) => {
              const { dependencies } = themeModule;

              const relevantDependencies = usedCssModules
                .map((moduleInfo) => ({
                  ...moduleInfo,
                  dependency: dependencies.find((d) => {
                    const module = compat.getDependencyModule(compilation, d);

                    return (
                      module && module.identifier() === moduleInfo.identifier
                    );
                  }),
                }))
                .filter(({ dependency }) => dependency);

              // Corrects Dependency.sourceOrder
              // Used for regular import chunk sorting, e.g. style-loader
              reIndexModules(
                relevantDependencies,
                {
                  getPreIndex: ({ dependency }) => dependency.sourceOrder,
                  getOwnerIndex: ({ owner }) =>
                    compat.getModulePreOrderIndex(compilation, owner),
                  getThemeIndex: ({ themeModule }) =>
                    compat.getModulePreOrderIndex(compilation, themeModule),
                  setPreIndex: ({ dependency }, i) => {
                    dependency.sourceOrder = i;
                  },
                  // post index not required for dependencies
                  getPostIndex: () => {},
                  setPostIndex: () => {},
                },
                'Dependency.sourceOrder',
              );
            });
        } catch (e) {
          compilation.errors.push(new TreatError(e));
        }
      });
    });

    compiler.hooks.normalModuleFactory.tap(TWP, (nmf) => {
      nmf.hooks.afterResolve.tap(TWP, (result) => {
        if (this.store.getCSSResources().has(result.matchResource)) {
          result.settings = Object.assign({}, result.settings, {
            sideEffects: true,
          });
        }
      });
    });

    const optionDefaulter = makeOptionDefaulter(
      isProductionLikeMode(compiler.options),
    );

    compiler.options.module.rules.splice(
      0,
      0,
      {
        test: this.test,
        use: [
          {
            loader: require.resolve('treat/webpack-plugin/loader'),
            options: {
              ...this.loaderOptions,
              minify: optionDefaulter(this.minify, {
                dev: false,
                prod: true,
              }),
              localIdentName: optionDefaulter(this.localIdentName, {
                dev: '[name]-[local]_[hash:base64:5]',
                prod: '[hash:base64:5]',
              }),
              themeIdentName: optionDefaulter(this.themeIdentName, {
                dev: '_[name]-[local]_[hash:base64:4]',
                prod: '[hash:base64:4]',
              }),
              store: this.store,
              treatCompiler: this.treatCompiler,
            },
          },
        ],
      },
      {
        test: /\.treatcss$/,
        sideEffects: true,
        use: [
          ...this.loaderOptions.outputLoaders,
          {
            loader: 'css-loader',
            options: {
              modules: false,
              url: false,
            },
          },
        ],
      },
    );
  }
}
