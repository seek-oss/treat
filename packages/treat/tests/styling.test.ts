import MiniCssExtractPlugin from 'mini-css-extract-plugin';
import startFixture from 'treat-test-helpers/startServer';
import getStyles from 'treat-test-helpers/getStyles';
import { TreatPlugin } from 'treat/webpack-plugin';

const mainSelector = '#main';

const configVariations = [
  {
    label: 'style-loader',
    config: () => ({
      plugins: [new TreatPlugin()],
    }),
  },
  {
    label: 'mini-css-extract-loader',
    config: () => ({
      optimization: {
        splitChunks: {
          cacheGroups: {
            styles: {
              name: 'styles',
              type: 'css/mini-extract',
              chunks: 'all',
              enforce: true,
            },
          },
        },
      },
      plugins: [
        new TreatPlugin({
          outputLoaders: [MiniCssExtractPlugin.loader],
        }),
        new MiniCssExtractPlugin(),
      ],
    }),
  },
  {
    label: 'development mode (style-loader)',
    config: () => ({
      mode: 'development',
      plugins: [new TreatPlugin()],
    }),
  },
];

describe('Styling and specificity', () => {
  configVariations.forEach(({ label, config }) => {
    describe(`Simple - ${label} - `, () => {
      let server: any;

      beforeAll(async () => {
        server = await startFixture({
          entry: require.resolve('./fixtures/simple/index.ts'),
          ...config(),
        });
        await page.goto(server.url);
      });

      describe('on mobile', () => {
        beforeAll(async () => {
          await page.setViewport({ width: 400, height: 800 });
        });

        it('should be styled correctly', async () => {
          const styles = await getStyles(page, mainSelector);

          expect(styles).toMatchSnapshot();
        });
      });

      describe('on desktop', () => {
        beforeAll(async () => {
          await page.setViewport({ width: 1200, height: 900 });
        });

        it('should be styled correctly', async () => {
          const styles = await getStyles(page, mainSelector);

          expect(styles).toMatchSnapshot();
        });

        describe('and hovered', () => {
          beforeAll(async () => {
            await page.hover(mainSelector);
          });

          it('should be styled correctly', async () => {
            const styles = await getStyles(page, mainSelector);

            expect(styles).toMatchSnapshot();
          });
        });
      });

      afterAll(() => {
        server.close();
      });
    });

    describe(`Themed - ${label}`, () => {
      let server: any;

      beforeAll(async () => {
        server = await startFixture({
          entry: require.resolve('./fixtures/themed/index.ts'),
          ...config(),
        });
        await page.goto(server.url);
      });

      it('should be styled correctly with theme one', async () => {
        const styles = await getStyles(page, '#theme1');

        expect(styles).toMatchSnapshot();
      });

      it('should be styled correctly with theme two', async () => {
        const styles = await getStyles(page, '#theme2');

        expect(styles).toMatchSnapshot();
      });

      afterAll(() => {
        server.close();
      });
    });

    describe(`Dynamic imports - ${label}`, () => {
      let server: any;

      beforeAll(async () => {
        server = await startFixture({
          entry: require.resolve('./fixtures/dynamic-imports/index.ts'),
          ...config(),
        });
        await page.goto(server.url);
      });

      describe('on mobile', () => {
        beforeAll(async () => {
          await page.setViewport({ width: 400, height: 800 });
        });

        it('theme A should be styled correctly', async () => {
          const styles = await getStyles(page, '#themeA');

          expect(styles).toMatchSnapshot();
        });

        it('theme B should be styled correctly', async () => {
          const styles = await getStyles(page, '#themeB');

          expect(styles).toMatchSnapshot();
        });
      });

      afterAll(() => {
        server.close();
      });
    });

    describe(`Unused modules - ${label}`, () => {
      let server: any;

      beforeAll(async () => {
        server = await startFixture({
          entry: require.resolve('./fixtures/unused-modules/index.ts'),
          ...config(),
        });
        await page.goto(server.url);
      });

      it('should be styled correctly', async () => {
        const styles = await getStyles(page, mainSelector);

        expect(styles).toMatchSnapshot();
      });

      afterAll(() => {
        server.close();
      });
    });
  });
});
