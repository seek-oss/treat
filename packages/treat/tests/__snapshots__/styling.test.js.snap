// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Styling and specificity Dynamic imports - development mode (style-loader) on mobile theme A should be styled correctly 1`] = `
Object {
  "background-color": "yellow",
  "display": "flex",
  "height": "100px",
  "width": "100px",
}
`;

exports[`Styling and specificity Dynamic imports - development mode (style-loader) on mobile theme B should be styled correctly 1`] = `
Object {
  "background-color": "red",
  "display": "flex",
  "height": "100px",
  "width": "100px",
}
`;

exports[`Styling and specificity Dynamic imports - mini-css-extract-loader on mobile theme A should be styled correctly 1`] = `
Object {
  "background-color": "#ff0",
  "display": "flex",
  "height": "100px",
  "width": "100px",
}
`;

exports[`Styling and specificity Dynamic imports - mini-css-extract-loader on mobile theme B should be styled correctly 1`] = `
Object {
  "background-color": "red",
  "display": "flex",
  "height": "100px",
  "width": "100px",
}
`;

exports[`Styling and specificity Dynamic imports - style-loader on mobile theme A should be styled correctly 1`] = `
Object {
  "background-color": "#ff0",
  "display": "flex",
  "height": "100px",
  "width": "100px",
}
`;

exports[`Styling and specificity Dynamic imports - style-loader on mobile theme B should be styled correctly 1`] = `
Object {
  "background-color": "red",
  "display": "flex",
  "height": "100px",
  "width": "100px",
}
`;

exports[`Styling and specificity Simple - development mode (style-loader) -  on desktop and hovered should be styled correctly 1`] = `
Object {
  "background-color": "purple",
  "color": "green",
  "display": "flex",
  "height": "100px",
  "position": "absolute",
  "width": "100px",
  "z-index": "2",
}
`;

exports[`Styling and specificity Simple - development mode (style-loader) -  on desktop should be styled correctly 1`] = `
Object {
  "color": "green",
  "display": "flex",
  "height": "100px",
  "position": "absolute",
  "width": "100px",
  "z-index": "2",
}
`;

exports[`Styling and specificity Simple - development mode (style-loader) -  on mobile should be styled correctly 1`] = `
Object {
  "color": "red",
  "display": "flex",
  "height": "100px",
  "position": "absolute",
  "width": "100px",
  "z-index": "2",
}
`;

exports[`Styling and specificity Simple - mini-css-extract-loader -  on desktop and hovered should be styled correctly 1`] = `
Object {
  "background-color": "purple",
  "color": "green",
  "display": "flex",
  "height": "100px",
  "position": "absolute",
  "width": "100px",
  "z-index": "2",
}
`;

exports[`Styling and specificity Simple - mini-css-extract-loader -  on desktop should be styled correctly 1`] = `
Object {
  "color": "green",
  "display": "flex",
  "height": "100px",
  "position": "absolute",
  "width": "100px",
  "z-index": "2",
}
`;

exports[`Styling and specificity Simple - mini-css-extract-loader -  on mobile should be styled correctly 1`] = `
Object {
  "color": "red",
  "display": "flex",
  "height": "100px",
  "position": "absolute",
  "width": "100px",
  "z-index": "2",
}
`;

exports[`Styling and specificity Simple - style-loader -  on desktop and hovered should be styled correctly 1`] = `
Object {
  "background-color": "purple",
  "color": "green",
  "display": "flex",
  "height": "100px",
  "position": "absolute",
  "width": "100px",
  "z-index": "2",
}
`;

exports[`Styling and specificity Simple - style-loader -  on desktop should be styled correctly 1`] = `
Object {
  "color": "green",
  "display": "flex",
  "height": "100px",
  "position": "absolute",
  "width": "100px",
  "z-index": "2",
}
`;

exports[`Styling and specificity Simple - style-loader -  on mobile should be styled correctly 1`] = `
Object {
  "color": "red",
  "display": "flex",
  "height": "100px",
  "position": "absolute",
  "width": "100px",
  "z-index": "2",
}
`;

exports[`Styling and specificity Themed - development mode (style-loader) should be styled correctly with theme one 1`] = `
Object {
  "background-color": "yellow",
  "display": "flex",
  "height": "12px",
  "line-height": "6",
  "z-index": "3",
}
`;

exports[`Styling and specificity Themed - development mode (style-loader) should be styled correctly with theme two 1`] = `
Object {
  "background-color": "red",
  "display": "flex",
  "height": "8px",
  "line-height": "4",
  "z-index": "3",
}
`;

exports[`Styling and specificity Themed - mini-css-extract-loader should be styled correctly with theme one 1`] = `
Object {
  "background-color": "#ff0",
  "display": "flex",
  "height": "12px",
  "line-height": "6",
  "z-index": "3",
}
`;

exports[`Styling and specificity Themed - mini-css-extract-loader should be styled correctly with theme two 1`] = `
Object {
  "background-color": "red",
  "display": "flex",
  "height": "8px",
  "line-height": "4",
  "z-index": "3",
}
`;

exports[`Styling and specificity Themed - style-loader should be styled correctly with theme one 1`] = `
Object {
  "background-color": "#ff0",
  "display": "flex",
  "height": "12px",
  "line-height": "6",
  "z-index": "3",
}
`;

exports[`Styling and specificity Themed - style-loader should be styled correctly with theme two 1`] = `
Object {
  "background-color": "red",
  "display": "flex",
  "height": "8px",
  "line-height": "4",
  "z-index": "3",
}
`;
